src/Microsoft.Diagnostics.Monitoring.StartupHook/Eventing/AbstractMonitorEventSource.cs

```diff
@@ -0,0 +1,170 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics.Tracing;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace Microsoft.Diagnostics.Monitoring.StartupHook.Eventing
+{
+    internal abstract class AbstractMonitorEventSource : EventSource
+    {
+        // Arrays are expected to have a 16-bit field for the length of the array.
+        // The length of the array is the number of elements, not the number of bytes.
+        private const int ArrayLengthFieldSize = sizeof(short);
+
+        /// <summary>
+        /// Amount of time to wait before sending batches of event source events in order to
+        /// avoid real-time buffering issues in the runtime eventing infrastructure and the
+        /// trace event library event processor.
+        /// </summary>
+        /// <remarks>
+        /// See: https://github.com/dotnet/runtime/issues/76704
+        /// </remarks>
+        private static readonly TimeSpan EventSourceBufferAvoidanceTimeout = TimeSpan.FromMilliseconds(200);
+
+        private readonly Timer _flushEventsTimer;
+
+        // NOTE: Arrays with a non-"byte" element type are not supported well by in-proc EventListener
+        // when using self-describing event format. This format is used to easily support event pipe listening.
+        public AbstractMonitorEventSource()
+            : base(EventSourceSettings.EtwSelfDescribingEventFormat)
+        {
+            _flushEventsTimer = new Timer(FlushTimerTick);
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            _flushEventsTimer.Dispose();
+            base.Dispose(disposing);
+        }
+
+        // Abstract base classes cannot define events so make the derived class define the flush event.
+        protected abstract void Flush();
+
+
+        [NonEvent]
+        private void FlushTimerTick(object? state)
+        {
+            Flush();
+        }
+
+        [NonEvent]
+        private void RestartFlushingEventTimer()
+        {
+            // This will reset the timer to fire after the specified time period. If the timer is already
+            // started, it will be reset. If it already finished, it will be started again.
+            _flushEventsTimer.Change(EventSourceBufferAvoidanceTimeout, Timeout.InfiniteTimeSpan);
+        }
+
+        [NonEvent]
+        protected unsafe void WriteEventCore(int eventId)
```

Comment:
This should not be necessary; there's already an `EventSource.WriteEvent(int)` method that can be called by deriving classes.

Reply:
> I can rename these methods to `WriteEventWithFlushing` if that'd be more clear

That sounds good to me.